src/crypto/city.cpp

```diff
@@ -516,15 +509,13 @@ uint128_t CityHash128WithSeed(const char *s, size_t len, uint128_t seed) {
   // different 56-byte-to-8-byte hashes to get a 16-byte final result.
   x = HashLen16(x, v.first);
   y = HashLen16(y + z, w.first);
-  return ( uint128_t( HashLen16(x + v.second, w.second) + y ) << 64 )
-         + HashLen16(x + w.second, y + v.second);
+  return uint128( HashLen16(x + v.second, w.second) + y, HashLen16(x + w.second, y + v.second) );
 }
 
 uint128_t city_hash128(const char *s, size_t len) {
   return len >= 16 ?
-      CityHash128WithSeed(s + 16, len - 16,
-                          ( uint128_t( Fetch64(s) ) << 64 ) + Fetch64(s + 8) + k0) :
-      CityHash128WithSeed(s, len, ( uint128_t(k0) << 64 ) + k1);
+      CityHash128WithSeed( s + 16, len - 16, uint128( Fetch64(s), Fetch64(s + 8) + k0 ) ) :
```

Comment:
I guess `Fetch64(s + 8) + k0` won't overflow?

Reply:
It can (and since k0 is large it will, quite often) but I think that's deliberate.
(Actually this means that the intermediate version in a3e2410 was broken but 1bda001 fixed it.)

Reply:
The original code was
```
uint128_t city_hash128(const char *s, size_t len) {
  return len >= 16 ?
      CityHash128WithSeed(s + 16, len - 16,
                          ( uint128_t( Fetch64(s) ) << 64 ) + Fetch64(s + 8) + k0) :
      CityHash128WithSeed(s, len, ( uint128_t(k0) << 64 ) + k1);
}
```
New code is
```
uint128_t city_hash128(const char *s, size_t len) {
  return len >= 16 ?
      CityHash128WithSeed( s + 16, len - 16, uint128( Fetch64(s), Fetch64(s + 8) + k0 ) ) :
      CityHash128WithSeed( s, len, uint128( k0, k1 ) );
}
```
While `uint128(...)` is defined as 
`inline uint128_t uint128( const uint64_t hi, const uint64_t lo )`

If `Fetch64(s + 8) + k0` would overflow, it looks like the results would be different.

I think the correct code should be
```
CityHash128WithSeed( s + 16, len - 16, uint128( Fetch64(s), Fetch64(s + 8) ) + k0 )
```

Reply:
No, the original code was
```
      CityHash128WithSeed(s + 16, len - 16,
                          uint128(Fetch64(s), Fetch64(s + 8) + k0)) :
      CityHash128WithSeed(s, len, uint128(k0, k1));
```
which was (wrongly, thanks for noticing!) changed by a3e2410 into 
```
      CityHash128WithSeed(s + 16, len - 16,
                          ( uint128_t( Fetch64(s) ) << 64 ) + Fetch64(s + 8) + k0) :
      CityHash128WithSeed(s, len, ( uint128_t(k0) << 64 ) + k1);
```
and then, by 1bda001, turned into the (once again correct) new code.

Reply:
Oh. You're correct.